<a href="https://colab.research.google.com/github/aditya452007/RESUME-BACKEND-PYTHON/blob/main/dataset_clean.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
try:
    data = pd.read_csv('/content/Dataset.csv', engine='python', on_bad_lines='skip')
except Exception as e:
    print(f"An error occurred: {e}")

In [ ]:
data

Category,Text
0                           Accountant,"________________
1                                               EDUCATION
2         OMBA - Executive Leadership University of Texas
3                                               2016-2018
4       O Bachelor of Science in Accounting Richland C...
...                                                   ...
885977  Freecodecamp.com Full Stack Developer Certific...
885978                             Planned Certifications
885979                            CompTIA Server+ sk0-004
885980                          CompTIA Security+ SY0-401
885981  Project Management Institute - Agile Certitied...

[885982 rows x 1 columns]

No charts were generated by quickchart


In [ ]:
data.isnull().sum()

Category,Text    6
dtype: int64

In [ ]:
file_path = '/content/Dataset.txt'
num_lines_to_display = 10

try:
    with open(file_path, 'r') as f:
        for i, line in enumerate(f):
            if i < num_lines_to_display:
                print(line.strip())
            else:
                break
except FileNotFoundError:
    print(f"Error: The file '{file_path}' was not found.")
except Exception as e:
    print(f"An error occurred: {e}")

Category,Text
Accountant,"________________

EDUCATION
OMBA - Executive Leadership University of Texas
2016-2018
O Bachelor of Science in Accounting Richland College
2005-2008
TRAINING & CERTIFICATIONS
Certified Management Accountant (CMA)


In [ ]:
try:
    data = pd.read_csv('/content/Dataset.txt', sep=',', engine='python', on_bad_lines='skip')
    display(data.head())
except Exception as e:
    print(f"An error occurred: {e}")

Category                                               Text
0  Accountant  ________________\r\n\r\nEDUCATION\r\nOMBA - E...
1  Accountant  ________________\r\n\r\nHOWARD GERRARD\r\nAcc...
2  Accountant  ________________\r\n\r\nKevin Frank\r\nSENIOR...
3  Accountant  ________________\r\n\r\nPlace of birth\r\nNat...
4  Accountant  ________________\r\n\r\nStephen Greet, CPA\r\...

In [ ]:
data.isnull().sum()

Category    0
Text        0
dtype: int64

In [ ]:
count_underscore = data['Text'].str.contains('________________').sum()
print(f"Number of entries containing '________________': {count_underscore}")

Number of entries containing '________________': 3410


In [ ]:
data['Text'] = data['Text'].str.replace('________________', '')
display(data['Text'].head())

0    \r\n\r\nEDUCATION\r\nOMBA - Executive Leaders...
1    \r\n\r\nHOWARD GERRARD\r\nAccountant\r\nDeyjo...
2    \r\n\r\nKevin Frank\r\nSENIOR ACCOUNTANT\r\ni...
3    \r\n\r\nPlace of birth\r\nNationality\r\nOliv...
4    \r\n\r\nStephen Greet, CPA\r\nSenior Accounta...
Name: Text, dtype: object

In [ ]:
print("Sampled Text entries after cleaning:")
for i in range(5):
    print(f"--- Sample {i+1} ---")
    print(data['Text'].sample(1).iloc[0])
    print("\n")

Sampled Text entries after cleaning:
--- Sample 1 ---


E-mail: info@qwikresume.com
ROBERT SMITH
Oracle PL/SQL Developer
SUMMARY
Phone: (0123)-456-789
Oracle PL/SQL Developer having 10 years of programming experience, as a Sr. Oracle SQL, PL/SQL and ETL Developer in Analysis, Design, and Implementation of Business Applications using the Oracle Relational Database Management System (RDBMS). Experience in Data Migration across various databases as well as within the same database across different systems, Well-versed in different stages of the Software Development Life Cycle (SDLC).
SKILLS
Oracle IIG, Oracle PL/SQL, UNIX Shell Script, SQL Developer, Toad, Oracle Developer, IBM Rational Agile & Waterfall, PowerBuilder
WORK EXPERIENCE
Oracle PL/SQL Developer
ABC Corp-2010-January 2018
• Created database objects like tables, views, sequences, synonyms, indexes using Oracle tools like SQL PLus, SQL developer, and toad.
⚫ Enforced data integrity using integrity constraint and database trigge

# Task
Analyze the dataset from "/content/Dataset.txt" to determine its readiness for training an ML model, including handling missing values, processing text data, and assessing data characteristics.

## Review data structure and types

### Subtask:
Examine the columns, their data types, and the overall structure of the DataFrame.


**Reasoning**:
Display the column names, their data types, and the shape of the DataFrame to understand its structure.



In [ ]:
data.info()
display(data.head())
print(data.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3411 entries, 0 to 3410
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Category  3411 non-null   object
 1   Text      3411 non-null   object
dtypes: object(2)
memory usage: 53.4+ KB


Category                                               Text
0  Accountant  \r\n\r\nEDUCATION\r\nOMBA - Executive Leaders...
1  Accountant  \r\n\r\nHOWARD GERRARD\r\nAccountant\r\nDeyjo...
2  Accountant  \r\n\r\nKevin Frank\r\nSENIOR ACCOUNTANT\r\ni...
3  Accountant  \r\n\r\nPlace of birth\r\nNationality\r\nOliv...
4  Accountant  \r\n\r\nStephen Greet, CPA\r\nSenior Accounta...

(3411, 2)


## Handle missing values

### Subtask:
Although we've checked for standard nulls, re-evaluate if there are other forms of missing or "unknown" values (like the "________________" strings) that need to be addressed.


**Reasoning**:
Inspect the 'Text' column for other potential non-standard missing values or irrelevant patterns by sampling and printing entries.



In [ ]:
print("Sampling more Text entries to identify other non-standard missing values/patterns:")
for i in range(10):
    print(f"--- Sample {i+1} ---")
    print(data['Text'].sample(1).iloc[0])
    print("\n")

Sampling more Text entries to identify other non-standard missing values/patterns:
--- Sample 1 ---


Harrison Sanford
58744 Mazie Springs, Philadelphia, PA Phone: +1 (555) 154 3342
EXPERIENCE DEVOPS LEAD-STARBUCKS TECHNOLOGY
EDUCATION
SKILLS
Houston, TX
09/2014-present
• Support the platform development team and perform activities to resolve developer issues in a timely and accurate fashion
• Implements deep telemetry capabilities to provide ongoing success and performance metrics
• Knowledge in application development and supporting a development environment
• Conducts research and makes recommendations on standards, products, and services in support of the Unified Commerce Platform
• 3+ years utilizing configuration management solutions as a system administrator
• Implements and maintains system configurations and baselines to support secure application development software control best practices and standards
• Communicates clearly and concisely, both orally and in writing
DEVOPS 

**Reasoning**:
Based on the sampled entries, it appears that the character '' is a common non-informative pattern at the beginning of the text. I will remove this character and also remove leading/trailing whitespace and newline characters to clean the 'Text' column further.



In [ ]:
data['Text'] = data['Text'].str.replace('', '', regex=False)
data['Text'] = data['Text'].str.strip()
print("Sampled Text entries after removing '' and stripping whitespace:")
for i in range(5):
    print(f"--- Sample {i+1} ---")
    print(data['Text'].sample(1).iloc[0])
    print("\n")

Sampled Text entries after removing '' and stripping whitespace:
--- Sample 1 ---
PROFILE
Michelle Baker, Food and
Beverage Attendant
FLAGSTAFF, 94103, VERENIGDE STATEN
890-455-111
Seeking a Food and Beverage Attendant position at Disneyworld Florida where hands-on food & beverage service experience and exceptional hospitality skills will be fully utilized to make an immediate contribution.
DETAILS
1355 Market St
Flagstaff, 94103, Verenigde Staten
890-455-111
rozenboomchantal@gmail.com
DATE / PLACE OF BIRTH 05/10/1987
Scottsdale
NATIONALITY
EMPLOYMENT HISTORY
Food and Beverage Attendant, Billabong Restaurant Jun. 2013 - Present, Broomfield, CO
Worked under the immediate supervision of the School Food Service Manager. Performed routine manual tasks in the preparation and serving of food and in cleaning and maintaining food
service areas.
American
DRIVING LICENSE
Full
SKILLS
Communication
Accomplishments:
• Memorized the restaurant's wine stock and the meals they should accompany, leadi

## Analyze categorical data

### Subtask:
If the 'Category' column is the target variable or a feature, analyze its unique values and distribution. Consider encoding strategies if necessary.


**Reasoning**:
Analyze the unique values and their distribution in the 'Category' column to understand the data's categorization and potential class imbalances.



In [ ]:
unique_categories = data['Category'].unique()
print("Unique categories in the 'Category' column:")
print(unique_categories)

category_counts = data['Category'].value_counts()
print("\nDistribution of categories in the 'Category' column:")
print(category_counts)

# If the number of unique categories is large, display only the top N
if len(unique_categories) > 20:
    print("\nTop 20 most frequent categories:")
    print(category_counts.head(20))

Unique categories in the 'Category' column:
['Accountant' 'Advocate' 'Agriculture' 'Apparel' 'Architecture' 'Arts'
 'Automobile' 'Aviation' 'Banking' 'Blockchain' 'BPO'
 'Building and Construction' 'Business Analyst' 'Civil Engineer'
 'Consultant' 'Data Science' 'Database' 'Designing' 'DevOps'
 'Digital Media' 'DotNet Developer' 'Education' 'Electrical Engineering'
 'ETL Developer' 'Finance' 'Food and Beverages' 'Health and Fitness'
 'Human Resources' 'Information Technology' 'Java Developer' 'Management'
 'Mechanical Engineer' 'Network Security Engineer' 'Operations Manager'
 'PMO' 'Public Relations' 'Python Developer' 'React Developer' 'Sales'
 'SAP Developer' 'SQL Developer' 'Testing' 'Web Designing']

Distribution of categories in the 'Category' column:
Category
Accountant                   151
Agriculture                  146
Architecture                 144
Aviation                     142
Advocate                     141
Arts                         138
Apparel                  

## Process text data

### Subtask:
The 'Text' column is in raw form. This step will involve cleaning the text (removing special characters, punctuation, etc.) and potentially transforming it into numerical features using techniques like TF-IDF or word embeddings.


**Reasoning**:
Define a text cleaning function, apply it to the 'Text' column, and then use TF-IDF vectorization to transform the cleaned text into numerical features.



In [ ]:
import re
from sklearn.feature_extraction.text import TfidfVectorizer

def clean_text(text):
    """
    Cleans the input text by converting to lowercase, removing punctuation,
    special characters, and extra whitespace.
    """
    text = text.lower() # Convert to lowercase
    text = re.sub(r'[^a-z0-9\s]', '', text) # Remove punctuation and special characters
    text = re.sub(r'\s+', ' ', text).strip() # Remove extra whitespace
    return text

# Apply the cleaning function
data['cleaned_text'] = data['Text'].apply(clean_text)

# Explore TF-IDF vectorization
# Initialize TF-IDF Vectorizer
# max_features can be adjusted based on computational resources and desired complexity
tfidf_vectorizer = TfidfVectorizer(max_features=5000, stop_words='english')

# Apply TF-IDF to the cleaned text
tfidf_matrix = tfidf_vectorizer.fit_transform(data['cleaned_text'])

print("Shape of the TF-IDF matrix:", tfidf_matrix.shape)
print("\nSample of the TF-IDF matrix (first 5 rows, first 10 features):")
print(tfidf_matrix[:5, :10].toarray())

# Explanation of choice:
print("\nExplanation of Vectorization Choice:")
print("TF-IDF (Term Frequency-Inverse Document Frequency) was chosen because it reflects the importance of a word in a document relative to its importance across the entire corpus.")
print("It helps to downweight common words (like 'the', 'a') and give more importance to words that are specific to certain documents or categories.")
print("This is often effective for tasks like text classification where differentiating documents based on their unique vocabulary is important.")
print("Considerations include the size of the vocabulary (controlled by max_features) and the potential need for stemming or lemmatization for further text normalization.")

Shape of the TF-IDF matrix: (3411, 5000)

Sample of the TF-IDF matrix (first 5 rows, first 10 features):
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]

Explanation of Vectorization Choice:
TF-IDF (Term Frequency-Inverse Document Frequency) was chosen because it reflects the importance of a word in a document relative to its importance across the entire corpus.
It helps to downweight common words (like 'the', 'a') and give more importance to words that are specific to certain documents or categories.
This is often effective for tasks like text classification where differentiating documents based on their unique vocabulary is important.
Considerations include the size of the vocabulary (controlled by max_features) and the potential need for stemming or lemmatization for further text normalization.


## Feature engineering

### Subtask:
Based on the problem you want to solve, consider creating new features from the existing data that could be useful for the ML model.


**Reasoning**:
Based on the problem of classifying resume categories, features like the length of the cleaned text or the word count could provide additional information to the model. I will create these two features.



In [ ]:
data['text_length'] = data['cleaned_text'].apply(len)
data['word_count'] = data['cleaned_text'].apply(lambda x: len(x.split()))
display(data.head())

Category                                               Text  \
0  Accountant  EDUCATION\r\nOMBA - Executive Leadership Unive...   
1  Accountant  HOWARD GERRARD\r\nAccountant\r\nDeyjob.com, Bi...   
2  Accountant  Kevin Frank\r\nSENIOR ACCOUNTANT\r\ninfo@resum...   
3  Accountant  Place of birth\r\nNationality\r\nOlivia Ogilvy...   
4  Accountant  Stephen Greet, CPA\r\nSenior Accountant\r\nThr...   

                                        cleaned_text  text_length  word_count  
0  education omba executive leadership university...         1449         188  
1  howard gerrard accountant deyjobcom birmingham...         2388         327  
2  kevin frank senior accountant inforesumekraftc...         2695         393  
3  place of birth nationality olivia ogilvy accou...         2703         387  
4  stephen greet cpa senior accountant through my...         1878         267

## Check for imbalance

### Subtask:
If this is a classification task, check if the classes in the target variable are imbalanced.


**Reasoning**:
Calculate and print the value counts of the 'Category' column to assess class distribution and discuss potential imbalance.



In [ ]:
category_counts = data['Category'].value_counts()
print("Distribution of categories in the 'Category' column:")
print(category_counts)

print("\nDiscussion on Class Imbalance:")
if category_counts.min() < category_counts.max() / 5: # A simple heuristic for potential imbalance
    print("The distribution of categories shows potential class imbalance, as some categories have significantly fewer samples than others.")
    print("For a classification task, this imbalance might need to be addressed during model training using techniques like oversampling, undersampling, or using appropriate evaluation metrics (e.g., F1-score, precision, recall).")
else:
    print("The distribution of categories appears relatively balanced.")


Distribution of categories in the 'Category' column:
Category
Accountant                   151
Agriculture                  146
Architecture                 144
Aviation                     142
Advocate                     141
Arts                         138
Apparel                      120
Banking                      118
Automobile                   113
Mechanical Engineer          103
Sales                         90
Civil Engineer                88
DevOps                        87
Human Resources               85
SQL Developer                 82
Electrical Engineering        79
Java Developer                78
Testing                       76
Operations Manager            76
Health and Fitness            76
Database                      71
Consultant                    70
Digital Media                 68
Public Relations              66
ETL Developer                 64
SAP Developer                 64
Business Analyst              64
DotNet Developer              63
Network Securi

## Assess data scale and distribution

### Subtask:
Understand the range and distribution of numerical features (if any are created). Scaling might be necessary for some models.


**Reasoning**:
Select the numerical columns created during feature engineering, calculate and display their descriptive statistics to understand their range and distribution.



In [ ]:
numerical_features = ['text_length', 'word_count']

print("Descriptive statistics for numerical features:")
display(data[numerical_features].describe())

Descriptive statistics for numerical features:


text_length   word_count
count  3411.000000  3411.000000
mean   2264.200235   312.882146
std     675.454491    92.080990
min      52.000000     7.000000
25%    1831.000000   253.000000
50%    2218.000000   306.000000
75%    2656.000000   367.000000
max    5939.000000   740.000000

## Final assessment

### Subtask:
Based on the findings from the previous steps, determine if the dataset is ready for ML model training or if further preprocessing is needed.


## Summary:

### Data Analysis Key Findings

*   The dataset contains 3411 rows and 2 columns: 'Category' and 'Text'. Both columns are of 'object' data type.
*   Initial checks revealed no standard missing values (nulls) in the 'Category' and 'Text' columns.
*   Non-standard patterns like the character '' and excessive whitespace were found and removed from the 'Text' column.
*   The 'Category' column contains 43 unique values, with a varying distribution across categories, indicating potential class imbalance (e.g., 'Accountant' with 151 occurrences vs. 'React Developer' with 18 occurrences).
*   The 'Text' column was cleaned by converting to lowercase, removing punctuation and special characters, and standardizing whitespace.
*   The cleaned text was transformed into a numerical representation using TF-IDF vectorization, resulting in a matrix of shape (3411, 5000).
*   Two new numerical features, 'text\_length' and 'word\_count', were engineered from the cleaned text.
*   The numerical features 'text\_length' (range 52 to 5939) and 'word\_count' (range 7 to 740) have different scales and distributions.
*   There is potential class imbalance in the 'Category' column, which might require specific handling for classification tasks.
*   The numerical features have different scales, suggesting that feature scaling might be beneficial for certain machine learning models.

### Insights or Next Steps

*   Given the class imbalance in the target variable ('Category') and the differences in scale for numerical features, further preprocessing steps like handling imbalance (e.g., oversampling, undersampling) and feature scaling (e.g., standardization, normalization) should be considered before training a machine learning model, especially for models sensitive to these issues.
*   The dataset is largely prepared for training a text classification model, but the identified issues (class imbalance, feature scaling) require attention. The next crucial step is to split the data into training and testing sets to evaluate model performance accurately.


In [ ]:
# Save the cleaned DataFrame to a CSV file
cleaned_file_path = '/content/cleaned_dataset.csv'
data.to_csv(cleaned_file_path, index=False)

print(f"Cleaned dataset saved to: {cleaned_file_path}")

Cleaned dataset saved to: /content/cleaned_dataset.csv


You can download the cleaned dataset using the file browser on the left pane in Colab.